semgrep/tests/performance/test_public_repos_perf.py

```diff
@@ -0,0 +1,47 @@
+import subprocess
+from dataclasses import dataclass
+from pathlib import Path
+
+import pytest
+
+
+@dataclass
+class RepoCase:
+    url: str
+    sha: str
+    language: str
+
+    @property
+    def test_id(self):
+        return self.url.split("/")[-1]
+
+
+@pytest.mark.parametrize(
+    "repo_case",
+    [
+        # SHAs of May 6, 2020
+        RepoCase("https://github.com/getsentry/sentry", "f25ea5dc", "python"),
+        RepoCase("https://github.com/highcharts/highcharts", "04709689", "javascript"),
+    ],
+    ids=lambda case: case.test_id,
+)
+def test_public_repos(run_semgrep_in_tmp, benchmark, repo_case):
+    subprocess.check_output(["git", "clone", repo_case.url, "repo"])
+    subprocess.check_output(
+        ["git", "--git-dir", "repo/.git", "checkout", repo_case.sha]
+    )
+    benchmark(
```

Comment:
I'm still a little confused as to why we need `subprocess` here. Can't we call `semgrep_main.main` directly? That way we avoid the `subprocess` overhead when benchmarking which could throw results off. Am I missing something?

Reply:
Two reasons:

1. The testing code is much simpler if we can reuse what we have in E2E tests, and we don't have to mock the `argparse.Namespace`.
2. I think it's fair to make benchmarks test the speed exactly how users experience it. Users will have to wait for the python process to spawn and modules to be imported. If we import semgrep in the test file before starting to benchmark, that'd hide issues such as https://github.com/returntocorp/semgrep/issues/633.

I was curious and checked how much time and variance the subprocess call adds, and I think it's not even that bad:

```
$ time python -c "import subprocess; import json; json.dumps([{'a':1} for _ in range(1000000)])"
0.59s user 0.12s system 98% cpu 0.790 total
$ time python -c "import subprocess; import json; json.dumps([{'a':1} for _ in range(1000000)])"
0.59s user 0.12s system 98% cpu 0.771 total
$ time python -c "import subprocess; import json; json.dumps([{'a':1} for _ in range(1000000)])"
0.59s user 0.12s system 98% cpu 0.771 total
$ time python -c "import subprocess; subprocess.check_output(['python', '-c', '''import json; json.dumps([{'a':1} for _ in range(1000000)])'''])"
0.61s user 0.13s system 98% cpu 0.804 total
$ time python -c "import subprocess; subprocess.check_output(['python', '-c', '''import json; json.dumps([{'a':1} for _ in range(1000000)])'''])"
0.62s user 0.13s system 98% cpu 0.803 total
$ time python -c "import subprocess; subprocess.check_output(['python', '-c', '''import json; json.dumps([{'a':1} for _ in range(1000000)])'''])"
0.62s user 0.13s system 98% cpu 0.837 total
```